modules/programs/fish.nix

```diff
@@ -0,0 +1,149 @@
+{ config, lib, pkgs, ...}:
+
+with lib;
+
+let
+
+  cfge = config.environment;
+
+  cfg = config.programs.fish;
+
+  fishVariables =
+    mapAttrsToList (n: v: ''set -x ${n} "${v}"'') cfg.variables;
+
+  shell = pkgs.runCommand pkgs.fish.name
+    { buildInputs = [ pkgs.makeWrapper ]; }
+    ''
+      source $stdenv/setup
+
+      mkdir -p $out/bin
+      makeWrapper ${pkgs.fish}/bin/fish $out/bin/fish
+    '';
+
+  fishAliases = concatStringsSep "\n" (
+    mapAttrsFlatten (k: v: "alias ${k} '${v}'") cfg.shellAliases
+  );
+
+in
+
+{
+
+  options = {
+
+    programs.fish = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to configure fish as an interactive shell.
+        '';
+      };
+
+      variables = mkOption {
+        type = types.attrsOf (types.either types.str (types.listOf types.str));
+        default = {};
+        description = ''
+          A set of environment variables used in the global environment.
+          These variables will be set on shell initialisation.
+          The value of each variable can be either a string or a list of
+          strings.  The latter is concatenated, interspersed with colon
+          characters.
+        '';
+        apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else v);
+      };
+
+      shellAliases = mkOption {
+        type = types.attrs;
+        default = cfge.shellAliases;
+        description = ''
+          Set of aliases for fish shell. See <option>environment.shellAliases</option>
+          for an option format description.
+        '';
+      };
+
+      shellInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell Script code called during fish shell initialisation.
+        '';
+      };
+
+      loginShellInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell script code called during fish login shell initialisation.
+        '';
+      };
+
+      interactiveShellInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell script code called during interactive fish shell initialisation.
+        '';
+      };
+
+      promptInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell script code used to initialise fish prompt.
+        '';
+      };
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    environment.etc."fish/foreign-env/shellInit" = mkIf (hasAttr "shellInit" cfge)
```

Comment:
The trouble is [`nix-daemon` `extraInit`](https://github.com/LnL7/nix-darwin/blob/5a5ec9ac42bcc99a6ddc577704824af225b99bd4/modules/services/nix-daemon.nix#L32) is not valid `fish`.